sqlglot/schema.py

```diff
@@ -262,63 +301,80 @@ def _normalize(self, schema: t.Dict) -> t.Dict:
             columns = nested_get(schema, *zip(keys, keys))
             assert columns is not None
 
-            normalized_keys = [self._normalize_name(key) for key in keys]
+            normalized_keys = [self._normalize_name(key, dialect=self.dialect) for key in keys]
             for column_name, column_type in columns.items():
                 nested_set(
                     normalized_mapping,
-                    normalized_keys + [self._normalize_name(column_name)],
+                    normalized_keys + [self._normalize_name(column_name, dialect=self.dialect)],
                     column_type,
                 )
 
         return normalized_mapping
 
-    def _normalize_table(self, table: exp.Table) -> exp.Table:
+    def _normalize_table(self, table: exp.Table, dialect: DialectType = None) -> exp.Table:
         normalized_table = table.copy()
+
         for arg in TABLE_ARGS:
             value = normalized_table.args.get(arg)
             if isinstance(value, (str, exp.Identifier)):
-                normalized_table.set(arg, self._normalize_name(value))
+                normalized_table.set(arg, self._normalize_name(value, dialect=dialect))
 
         return normalized_table
 
-    def _normalize_name(self, name: str | exp.Identifier) -> str:
+    def _normalize_name(self, name: str | exp.Identifier, dialect: DialectType = None) -> str:
+        dialect = dialect or self.dialect
+
         try:
-            identifier = sqlglot.maybe_parse(name, dialect=self.dialect, into=exp.Identifier)
+            identifier = sqlglot.maybe_parse(name, dialect=dialect, into=exp.Identifier)
         except ParseError:
             return name if isinstance(name, str) else name.name
 
-        return identifier.name if identifier.quoted else identifier.name.lower()
+        name = identifier.name
+
+        if identifier.quoted:
+            return name
+
+        # Only Snowflake resolves unquoted identifiers as uppercase.
+        # https://docs.snowflake.com/en/sql-reference/identifiers-syntax
+        return name.upper() if dialect == "snowflake" else name.lower()
```

Comment:
maybe we should make this a dialect flag, there could be others in the future, and i don't like how snowflake is hardcoded in 2 places now

Reply:
Agreed, will try to refactor

Reply:
Took a stab at this, let me know what you think. Happy to clean up if needed.